File: src/data/codes/linear-svm.m
clc;
clear;
close all;

%% =========================
% SETTINGS
% ==========================
useSynthetic = true;
csvFile = 'linear_svm_placement.csv';

%% =========================
% LOAD OR CREATE DATA
% ==========================
if useSynthetic
    rng(11);
    n = 140;

    CGPA = 5 + 5*rand(n,1);
    Aptitude_Score = randi([40 100], n, 1);
    Coding_Score = randi([35 100], n, 1);
    Internship = randi([0 1], n, 1);

    score = 1.2*CGPA + 0.03*Aptitude_Score + 0.035*Coding_Score + 0.8*Internship;
    Placement = double(score > 11);

    data = table(CGPA, Aptitude_Score, Coding_Score, Internship, Placement);
else
    data = readtable(csvFile);
    % Expected columns:
    % CGPA, Aptitude_Score, Coding_Score, Internship, Placement
end

disp(head(data));

%% =========================
% PREPROCESSING
% ==========================
data = rmmissing(data);
data = unique(data);

X = [data.CGPA, data.Aptitude_Score, data.Coding_Score, data.Internship];
Y = data.Placement;

X = normalize(X);

%% =========================
% TRAIN-TEST SPLIT
% ==========================
cv = cvpartition(Y, 'HoldOut', 0.2);
X_train = X(training(cv), :);
Y_train = Y(training(cv), :);
X_test  = X(test(cv), :);
Y_test  = Y(test(cv), :);

%% =========================
% MODEL BUILDING
% ==========================
model = fitcsvm(X_train, Y_train, 'KernelFunction', 'linear', 'Standardize', true);

%% =========================
% PREDICTION
% ==========================
Y_pred = predict(model, X_test);

%% =========================
% EVALUATION
% ==========================
confMat = confusionmat(Y_test, Y_pred);
accuracy = sum(diag(confMat)) / sum(confMat(:));

TP = confMat(2,2); FP = confMat(1,2); FN = confMat(2,1);
precision = TP / (TP + FP + eps);
recall    = TP / (TP + FN + eps);
f1        = 2 * precision * recall / (precision + recall + eps);

disp('Confusion Matrix:');
disp(confMat);
fprintf('\nAccuracy  = %.4f\n', accuracy);
fprintf('Precision = %.4f\n', precision);
fprintf('Recall    = %.4f\n', recall);
fprintf('F1-score  = %.4f\n', f1);

figure;
confusionchart(Y_test, Y_pred);
title('Linear SVM Confusion Matrix');

File: src/data/codes/linear-svm.py
"""Linear-kernel SVM
# pip install numpy pandas scikit-learn matplotlib
"""
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

USE_SYNTHETIC = True
RNG = np.random.default_rng(11)

if USE_SYNTHETIC:
    n = 140
    cgpa = RNG.uniform(5, 10, size=n)
    apt = RNG.integers(40, 101, size=n)
    code = RNG.integers(35, 101, size=n)
    intern = RNG.integers(0, 2, size=n)
    score = 1.2 * cgpa + 0.03 * apt + 0.035 * code + 0.8 * intern
    placement = (score > 11).astype(int)
    data = pd.DataFrame(
        {
            "CGPA": cgpa,
            "Aptitude_Score": apt,
            "Coding_Score": code,
            "Internship": intern,
            "Placement": placement,
        }
    )
else:
    data = pd.read_csv("linear_svm_placement.csv")

print(data.head())
data = data.dropna().drop_duplicates()
X = data[["CGPA", "Aptitude_Score", "Coding_Score", "Internship"]].values
y = data["Placement"].values
scaler = StandardScaler()
X = scaler.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

model = SVC(kernel="linear", random_state=42).fit(X_train, y_train)
y_pred = model.predict(X_test)

cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", cm)
print(f"Accuracy  = {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision = {precision_score(y_test, y_pred, zero_division=0):.4f}")
print(f"Recall    = {recall_score(y_test, y_pred, zero_division=0):.4f}")
print(f"F1-score  = {f1_score(y_test, y_pred, zero_division=0):.4f}")

fig, ax = plt.subplots()
ConfusionMatrixDisplay.from_predictions(y_test, y_pred, ax=ax)
ax.set_title("Linear SVM Confusion Matrix")
plt.show()
